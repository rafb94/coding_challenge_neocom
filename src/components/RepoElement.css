.container{
    border-radius: 10px;
    margin: 2em 4em;
    background-color: #e6e6e6;
    height: 6em;
    line-height: 6em;
    border: 1px solid black;

    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
}

.starred{
    background-color:#ffb3b3;
}

.starred button{
    opacity: 0.5
}

.container>* {
    flex-grow: 3;
    flex-basis: 0;
    text-align: center;
}

button {
    height: 5em;
    width: 10em;
    border-radius: 5px;
}