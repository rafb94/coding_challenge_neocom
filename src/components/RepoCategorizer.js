// Own imports
import RepoElement from './RepoElement.js';


function RepoCategorizer({repos, message, setRepos}) {

    let addStarsToRepo = (specifiedRepo, starCount, starred) =>
    {
        let newRepos = repos.map(
            (repo) =>
            {
                if (repo.id === specifiedRepo.id)
                {
                    let newRepo = {
                        ...repo, 
                        stargazers_count: repo.stargazers_count + starCount, 
                        starred: starred
                    };

                    return newRepo
                }

                return repo;
            }
        )

        setRepos(newRepos);
    }


    let filterRepos = (lowerLimit, upperLimit) =>
    {
        let returnValue = repos.filter(
            (repo) =>
            {
                // conditionally filter when both the lower and upper limit are defined
                if (repo.stargazers_count >= lowerLimit && upperLimit && repo.stargazers_count <= upperLimit)
                {
                    return true;
                }
                // conditionally filter when only the lower limit is defined
                else if (repo.stargazers_count >= lowerLimit && !upperLimit)
                {
                    return true;
                }

                return false;
            }
        ) 

        return returnValue;
    }

    let createRepoElements = (repos) =>
    {
        let returnValue = repos.map(
            (repo) =>
            {
                return <RepoElement key={repo.id} repo={repo} addStarsToRepo={addStarsToRepo} />;
            }
        )

        return returnValue;
    }

    return (
        <div>
            <div>
                <h1> More than 10.000 stars</h1>
                {
                    repos !== undefined? 
                    createRepoElements(filterRepos(10001)):
                    message
                }
            </div>

            <div>
                <h1> 1.000 to 10.000 stars</h1>
                {
                    repos !== undefined? 
                    createRepoElements(filterRepos(1000, 10000)):
                    message
                }
            </div>

            <div>
                <h1> Below 1.000 stars</h1>
                {
                    repos !== undefined? 
                    createRepoElements(filterRepos(0, 999)):
                    message
                }
            </div>
            
        </div>
    );
}

export default RepoCategorizer;