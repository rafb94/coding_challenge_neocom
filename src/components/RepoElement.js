// External imports
import {useState} from 'react';

// Own imports
import './RepoElement.css';

function RepoElement({repo, addStarsToRepo}) {

    // use info about the "starred" status if available, otherwise the repo is not "starred"
    let [starred, setStarred] = useState(repo && repo.starred? repo.starred: false);

    let toggleStarred = () =>
    {
        let newStarred = !starred;
        setStarred(newStarred);

        let starCountToBeAdded = 0;
        if (starred === true)
        {
            starCountToBeAdded = -1000;
        }
        else
        {
            starCountToBeAdded = 1000;
        }
        
        addStarsToRepo(repo, starCountToBeAdded, newStarred);
    }

    return (
        <div className={`container${starred? " starred": ""}`} >
            <div>
                <span><a href={repo.html_url}>{repo.name? repo.name: "No name available."}</a></span>
            </div>

            <div>
                <span>{repo.stargazers_count? repo.stargazers_count: "No star info available."}</span>
            </div>

            <div>
                {
                    starred?
                    <button onClick={toggleStarred}>
                        <span>unstar</span>
                    </button>
                    :
                    <button onClick={toggleStarred}>
                        <span>star</span>
                    </button>
                }
            </div>
        </div>
    );
}

export default RepoElement;