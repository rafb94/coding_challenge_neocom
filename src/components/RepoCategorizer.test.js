// External imports
import {describe, expect, test} from '@jest/globals';
import {shallow} from 'enzyme';

// Own imports
import RepoCategorizer from './RepoCategorizer.js';
import RepoElement from './RepoElement.js';

describe("Test RepoCategorizer with 4 repos",
    () =>
    {
        let repos = [
            {
              "id": 132935648,
              "name": "repo123",
              "stargazers_count": 14,
            },
            {
              "id": 18221276,
              "name": "halloWelt",
              "stargazers_count": 20,
            },
            {
              "id": 20978623,
              "name": "hello-worId",
              "stargazers_count": 46,
            },
            {
                "id": 1000,
                "name": "hello-worId1",
                "stargazers_count": 10001,
            },
        ]

        const container = shallow(<RepoCategorizer repos={repos} />);

        it('should match snapshot', () => {
            expect(container).toMatchSnapshot();
        });

        it('should contain three h1 titles', () => {
            expect(container.find('h1').length).toEqual(3);
        });

        it('should contain four RepoElements', () => {
            expect(container.find(RepoElement).length).toEqual(4);
        });

        it('First category should contain one element', () => {
            expect(container.find('div').at(0).childAt(0).find(RepoElement).length).toEqual(1);
        });

        it('Second category should contain zero elements', () => {
            expect(container.find('div').at(0).childAt(1).find(RepoElement).length).toEqual(0);
        });

        it('Third category should contain three elements', () => {
            expect(container.find('div').at(0).childAt(2).find(RepoElement).length).toEqual(3);
        });
    }
)

describe("Test RepoCategorizer with no repos",
    () =>
    {
        const container = shallow(<RepoCategorizer />);

        it('should match snapshot', () => {
            expect(container).toMatchSnapshot();
        });

        it('should contain three h1 titles', () => {
            expect(container.find('h1').length).toEqual(3);
        });

        it('should contain zero RepoElements', () => {
            expect(container.find(RepoElement).length).toEqual(0);
        });
    }
)
