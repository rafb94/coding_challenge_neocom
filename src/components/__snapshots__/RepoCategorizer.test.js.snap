// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RepoCategorizer with 4 repos should match snapshot 1`] = `ShallowWrapper {}`;

exports[`Test RepoCategorizer with no repos should match snapshot 1`] = `ShallowWrapper {}`;
